docs/proposals/20210316-containerrecreaterequest.md

```diff
@@ -0,0 +1,108 @@
+---
+title: Proposal Template
+authors:
+  - "@FillZpp"
+reviewers:
+  - "@Fei-Guo"
+  - "@furykerry"
+creation-date: 2021-03-16
+last-updated: 2021-03-16
+status: implementable
+---
+
+# ContainerRecreateRequest
+
+- Provide a way to let users restart/recreate one or more containers in a Pod.
+
+## Table of Contents
+
+A table of contents is helpful for quickly jumping to sections of a proposal and for highlighting
+any additional information provided beyond the standard proposal template.
+[Tools for generating](https://github.com/ekalinin/github-markdown-toc) a table of contents from markdown are available.
+
+- [ContainerRecreateRequest](#containerrecreaterequest)
+  - [Table of Contents](#table-of-contents)
+  - [Motivation](#motivation)
+  - [Proposal](#proposal)
+    - [User Stories](#user-stories)
+      - [Story 1](#story-1)
+    - [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
+    - [Risks and Mitigations](#risks-and-mitigations)
+  - [Implementation History](#implementation-history)
+
+## Motivation
+
+Now that the only way to recreate a container in an active Pod is to update it in-place.
+However, it also requires users to change the image defined in workloads (e.g. CloneSet/Advanced StatefulSet).
+
+We hope to provide a way to let users recreate(restart) one or more containers in any active Pods.
+It is useful for some user scenarios, such as:
+
+1. Restart the application in app container.
+2. Restart a sidecar contaienr along with app container.
+
+## Proposal
+
+Define a CRD named `ContainerRecreateRequest` (short name `CRR`):
+
+```yaml
+apiVersion: apps.kruise.io/v1alpha1
+kind: ContainerRecreateRequest
+metadata:
+  namespace: pod-namespace
+  name: xxx
+  labels:
+    crr.apps.kruise.io/pod-name: pod-name
+    crr.apps.kruise.io/node-name: node-name-of-the-pod  # injected by kruise webhook
+spec:
+  containers:
+  - name: app
+  - name: sidecar
+    skipPreStop: false           # 'true' means do not execute preStop hook before stop the container
+  strategy:
+    failurePolicy: Fail          # or 'Ignore'
+    stopGracePeriodSeconds: 30   # the timeout period for preStop hook and container stopping
+    readyProtection: false       # 'true' means make the Pod not-ready before recreating containers
+    continueByReady: false       # 'true' means recreate the next container only if the previous one has been recreated and ready
+  activeDeadlineSeconds: 300        # deadline of this ContainerRecreateRequest
+  ttlSecondsAfterFinished: 1800     # ttl after this ContainerRecreateRequest has finished
```

Comment:
please add ContainerRecreateRequest status description, and describe how user determine the recreate is successful